software/contrib/preset_manager.py

```diff
@@ -0,0 +1,222 @@
+
+    def draw_preset_list(self):
+        oled.fill(0)
+
+        positions = len(self.presets) + 1
+        self.hovered_preset = k1.read_position(positions)
+
+        vertical_offset = -(self.hovered_preset * self.list_line_height) + 1
+        oled.fill_rect(0, 0, OLED_WIDTH, self.list_line_height, 1)
+        for index, preset in enumerate(self.presets):
+            name = preset[0]
+            oled.text(name, 0, vertical_offset, 1 if self.hovered_preset != index else 0)
+
+            vertical_offset += self.list_line_height
+        if self.hovered_preset == (positions - 1):
+            colour = 0
+            self.create_new_preset = True
+        else:
+            colour = 1
+            self.create_new_preset = False
+        oled.text("NEW PRESET", 0, vertical_offset, colour)
+        oled.show()
+
+    def load_new_preset(self):
+        if self.create_new_preset:
+            self.presets.append(self.create_blank_preset("blank preset"))
+            self.loaded_preset = -1
+        else:
+            self.loaded_preset = self.hovered_preset
+
+    def draw_rename_screen(self):
+        if not self.create_new_preset:
+            oled.fill(0)
+
+            name = self.presets[self.hovered_preset][0]
+            name_list = list(name)
+            while len(name_list) < self.max_characters:
+                name_list.append(" ")
+
+            old_selected_letter = self.selected_letter
+            self.selected_letter = k1.read_position(self.max_characters)
+            if old_selected_letter != self.selected_letter:
+                self.old_k2 = k2.read_position(len(self.characters) + 1)
+            else:
+                if k2.read_position(len(self.characters) + 1) != self.old_k2:
+                    name_list[self.selected_letter] = self.characters[
+                        k2.read_position(self.number_of_characters)
+                    ]
+                    self.presets[self.hovered_preset][0] = "".join(name_list)
+
+            oled.fill_rect((self.selected_letter * CHAR_WIDTH), 0, CHAR_WIDTH, (CHAR_HEIGHT + 2), 1)
+            for index, letter in enumerate(name_list):
+                oled.text(
+                    letter, (index * CHAR_WIDTH), 1, 0 if self.selected_letter == index else 1
+                )
+
+            oled.text("< DELETE", 0, 24, 1)
+        else:
+            self.screen = 0  # If a user tries to go to the rename screen while hovering over 'NEW PRESET' they are kept on the preset list screen
+
+    def hover_delete(self):
+        oled.fill(0)
+        oled.fill_rect(0, 22, 128, 32, 1)
+        oled.text("CANCEL >", 64, 0, 1)
+        oled.text("< CONFIRM DELETE?", 0, 24, 0)
+        oled.show()
+
+    def update_ui(self):
+        if self.confirm_delete:
+            if len(self.presets) > 1:
+                if self.hovered_preset == self.loaded_preset:
+                    self.loaded_preset = 0
+                self.presets.remove(self.presets[self.hovered_preset])
```

Comment:
[required] There's a bug here where the index in `loaded_preset` can become out of bounds. That is it points to an index that is no longer valid in the `presets` list. If this bad index gets saved to the state file then the script will crash on launch.

To reproduce this select the last preset, then delete one from earlier in the list.

```
Traceback (most recent call last):
  File "main.py", line 55, in <module>
  File "/lib/bootloader.py", line 90, in main
  File "/lib/contrib/preset_manager.py", line 216, in main
  File "/lib/contrib/preset_manager.py", line 211, in update_ui
  File "/lib/contrib/preset_manager.py", line 93, in draw_current_preset
IndexError: list index out of range
```

This should be solved in two ways. First the delete process should manage the loaded_preset index to keep it valid. Second, the save state method should ensure that it is writing a valid state, that is, it shouldn't write a 'loaded' index that is out of bounds.